lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp

```diff
@@ -1472,4 +1472,99 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
                       binder.op, resultType, operand);
                   return success();
                 });
+  patterns.onOp(
+      "ConstantOfShape", 20,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Torch::ValueTensorType resultType;
+        Value shape;
+        if (binder.tensorOperand(shape) || binder.tensorResultType(resultType))
+          return failure();
+
+        // convert shape tensor to list of ints
+        auto shapeSizes =
+            dyn_cast<Torch::ValueTensorType>(shape.getType()).getSizes();
+        SmallVector<Value> dimList;
+        SmallVector<int64_t> selectSizes;
+        selectSizes.push_back(1);
+        Torch::BaseTensorType shapeType =
+            shape.getType().cast<Torch::BaseTensorType>();
+        Type selectResultType = shapeType.getWithSizesAndDtype(
+            llvm::ArrayRef(selectSizes), shapeType.getOptionalDtype());
+        Value zero = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+        for (int i = 0; i < shapeSizes[0]; i++) {
+          Value selectIndex = rewriter.create<Torch::ConstantIntOp>(
+              binder.getLoc(), rewriter.getType<Torch::IntType>(),
+              rewriter.getIntegerAttr(rewriter.getIntegerType(64), i));
+          Value extract = rewriter.create<Torch::AtenSelectIntOp>(
+              binder.getLoc(), selectResultType, shape, zero, selectIndex);
+          Value dim = rewriter.create<Torch::AtenItemOp>(
+              binder.getLoc(), rewriter.getType<Torch::IntType>(), extract);
+          dimList.push_back(dim);
+        }
+        Value dimValueList = rewriter.create<Torch::PrimListConstructOp>(
+            binder.getLoc(),
+            Torch::ListType::get(Torch::IntType::get(binder.op->getContext())),
+            dimList);
+        Value noneVal = rewriter.create<Torch::ConstantNoneOp>(binder.getLoc());
+
+        // Get fill_value if it is present.
+        Value value_const;
+        auto attr = binder.op->getAttr("torch.onnx.value");
+        auto resultDType = resultType.getDtype();
+        if (!attr) {
+          if (!resultDType.isa<FloatType>()) {
+            return rewriter.notifyMatchFailure(
+                binder.op,
+                "`value` attr and result type (FloatType) mismatch.");
+          }
+          value_const = rewriter.create<Torch::ConstantFloatOp>(
+              binder.getLoc(), rewriter.getF64FloatAttr(0.0));
+        } else if (attr.isa<mlir::DenseElementsAttr>()) {
```

Comment:
We want to avoid all of these `if {} else if {} else if {} else{}` chains. Often we can know how the changes sequence and avoiding all the nesting / overlappiing.

For instance, your `if (!attr) { ...}` case can be seen as setting the default, then each other case (e.g. `if(attr.isa<DenseElementsAttr>()) {...}` can handle the individual cases. You can also do an upfront `if(!isa<Type1, Type2, Type3>(attr)( {return failure();}` to upfront validate the supported types.

Reply:
I agree those if else chains were ugly, thanks for this review, pushed to try and get rid of them.
Most of the nesting was to ensure that any mismatch in resultDType and value attr dtype shold fail right away but I realise I might have been overly cautious there. Proceeding with an assumption for now and will fix if that break later on. 
Kept only `DenseElementsAttr ` as acceptable type since ONNX specifies value attr to be a TENSOR type.